src/components/IntlTelInput.d.ts

```diff
@@ -0,0 +1,494 @@
+    extension: string,
+  ) => void
+  /**
+   * Optional validation callback function. It returns validation status, input box value and selected country data.
+   * @default null
+   */
+  onPhoneNumberBlur?: (
+    isValid: boolean,
+    value: string,
+    seletedCountryData: CountryData,
+    fullNumber: string,
+    extension: string,
+    event: React.FocusEvent<HTMLInputElement>,
+  ) => void
+  /**
+   * Optional validation callback function. It returns validation status, input box value and selected country data.
+   * @default null
+   */
+  onPhoneNumberFocus?: (
+    isValid: boolean,
+    value: string,
+    seletedCountryData: CountryData,
+    fullNumber: string,
+    extension: string,
+    event: React.FocusEvent<HTMLInputElement>,
+  ) => void
+  /**
+   * Allow main app to do things when a country is selected.
+   * @default null
+   */
+  onSelectFlag?: (
+    currentNumber: string,
+    seletedCountryData: CountryData,
+    fullNumber: string,
+    isValid: boolean,
+  ) => void
+  /**
+   * Disable this component.
+   * @default false
+   */
+  disabled?: boolean
+  /**
+   * Static placeholder for input controller. When defined it takes priority over autoPlaceholder.
+   */
+  placeholder?: string
+  /**
+   * Enable auto focus
+   * @default false
+   */
+  autoFocus?: boolean
+  /**
+   * Set the value of the autoComplete attribute on the input.
+   * For example, set it to phone to tell the browser where to auto complete phone numbers.
+   * @default 'off'
+   */
+  autoComplete?: string
+  /**
+   * Style object for the wrapper div. Useful for setting 100% width on the wrapper, etc.
+   */
+  style?: React.CSSProperties
+  /**
+   * Render fullscreen flag dropdown when mobile useragent is detected.
+   * The dropdown element is rendered as a direct child of document.body
+   * @default true
+   */
+  useMobileFullscreenDropdown?: boolean
+  /**
+   * Pass through arbitrary props to the tel input element.
+   * @default {}
+   */
+  telInputProps?: React.InputHTMLAttributes<HTMLInputElement>
+  /**
+   * Format the number.
+   * @default true
+   */
+  format?: boolean
+  /**
+   * Allow main app to do things when flag icon is clicked.
+   * @default null
+   */
+  onFlagClick?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void
+}
+
+export interface IntlTelInputState {
+  showDropdown: boolean
+  highlightedCountry: number
+  value: string
+  disabled: boolean
+  readonly: boolean
+  offsetTop: number
+  outerHeight: number
+  placeholder: string
+  title: string
+  countryCode: string
+  dialCode: string
+  cursorPosition: any
+}
+
+export default class IntlTelInput extends React.Component<
+  IntlTelInputProps,
+  IntlTelInputState
+> {
+  //#region Properties
+  wrapperClass: {
+    [key: string]: boolean
+  }
+
+  defaultCountry?: string
+
+  autoCountry: string
+
+  tempCountry: string
+
+  startedLoadingAutoCountry: boolean
+
+  dropdownContainer?: React.ElementType | ''
+
+  isOpening: boolean
```

Comment:
You're going to see a lot of me declaring every property and method that's accessible at the JS level.

I'm pretty sure that this won't make plenty of sense (in TypeScript, the best practice is to expose only properties and methods we intentionally declare `public`, then leave the rest out). I'm not quite certain whether we have API docs on methods we want people to use, so in doubt, I exposed everything.

We can still give this a follow-up PR after this one has been merged, of course.